### PR TITLE
ci: add markdown link checker

### LINKED ISSUE #731: Add markdown checker
As file structure and external URL changes, we might end up with some broken URLs on our markdown files. We should configure some markdown checker to minimize the need to perform this task manually.

### PR DESCRIPTION
**Pull request by Dhaiwat10**
## Summary
- This PR adds a new job called `validate-markdown-links` to our PR Lint workflow that scans *only the modified* markdown files in a PR for broken links.
- We could configure this job to scan for *all* files and not just the modified files but that would make the check extremely slow. Let me know if this is something we need, though!

## To-do
- [ ] Find a fix for the 'dynamic' links in our docs site's content. [Context](https://github.com/FuelLabs/fuels-ts/pull/737#issuecomment-1398269676)
- [ ] Create a script in our `package.json` that lets us run this check on _all_ the files in the codebase locally if we want.

Closes #731

### CODE REVIEW

**Comment by arboleya** [comment]
1. Does this validate external links as well?
1. Did you try running it against _all_ files to see how long it takes?

We should try validating all `.md` files, if possible. The ones standalone (readme, contributing, etc.) and the ones from the Docs/Guide. This would guarantee we'd know when some external link breaks, not only manual mistakes made on modified files. A good example is the links to the Sway docs.

**Comment by digorithm** [comment]
Needs to re-sync with master!

**Comment by Dhaiwat10** [comment]
@arboleya 

1. Yeah it does handle external links!
2. It was taking longer than *25 minutes* (it failed at 25 mins because it found a broken link) :/

Let me see if I can do something to cut down the running time. One easy win I can see is ignoring the files inside of `_guide` because the same content is then sent to the auto-generated `guide` folder anyways.

**Comment by arboleya** [comment]
Alright, that is way too slow. Your first approach is the best, indeed.

> _In the future, we could search for a similar tool that we can run locally and on-demand, to scan all files._

**Comment by Dhaiwat10** [comment]
> In the future, we could search for a similar tool that we can run locally and on-demand, to scan all files.

@arboleya I think we can use this command for that:

```bash
find . -name \*.md -print0 | xargs -0 -n1 markdown-link-check
```

We'll need to use this package called [`markdown-link-check`](https://github.com/tcort/markdown-link-check#installation). I think the Github action I added uses the same package under the hood.

I can add this as a script in our package.json.

**Comment by arboleya** [comment]
@Dhaiwat10 Perfect!

> **Note** _CI (PR title validation) seems frozen; I tried renaming it but it didn't help. A new commit might do it._

**Comment by Dhaiwat10** [comment]
Fixed the formatting! Having some issues with the command to run the checks locally, though. Figuring it out

**Comment by Dhaiwat10** [comment]
I came across a bug: with the current setup this action won't be able to parse 'dynamic' links like this one in our docs site's content. Need to find a workaround for this, too.

image

**Comment by arboleya** [comment]
@camsjams Would it make sense to run this validation against the `guide` directory instead?

Any drawbacks?

**Comment by Dhaiwat10** [comment]
@camsjams that sounds fair, let me take a stab at this

**Comment by Dhaiwat10** [comment]
This is how I'm thinking of going ahead with this one, please correct me if I'm wrong:

- We will have a Github workflow that checks _only_ the modified files for broken links (this is buggy rn, working on a fix)
- We will have a local / npm script (eg. `lint:docs`) that, when ran, scans our _entire_ codebase for broken links in markdown files. (right now it _only_ scans the docs)

Have made some decent progress but still need to iron out some edge cases that keep popping up.

**Comment by Dhaiwat10** [comment]
This is how I'm thinking of going ahead with this one, please correct me if I'm wrong:

- We will have a Github workflow that checks _only_ the modified files for broken links (this is buggy rn, working on a fix)
- We will have a local / npm script (eg. `lint:docs`) that, when ran, scans our _entire_ codebase for broken links in markdown files. (right now it _only_ scans the docs)

Have made some decent progress but still need to iron out some edge cases that keep popping up.

File: .github/workflows/pr-lint.yaml [hunk 1]
```diff
@@ -35,3 +35,14 @@ jobs:
...
+  validate-markdown-links:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@master
+
+      - name: CI Setup
+        uses: ./.github/actions/ci-setup
+
+      - name: Lint docs
+        run: pnpm lint:docs
```
Comment (Dhaiwat10) [inline .github/workflows/pr-lint.yaml:48]: Running `lint:docs` instead of the previous action I was using to temporarily debug something. Will revert back to the action that only scans modified files.

**Comment by Dhaiwat10** [comment]
Converting this to a draft until it's ready to be reviewed.

**Comment by Dhaiwat10** [comment]
An update on this PR:

1. I think we should reduce the scope of this PR to _only_ adding a Github action that checks the modified files in a PR for now. I propose creating a separate issue for a local npm script that runs the check on _all_ the files in our codebase. There might be some additional work needed for that.
2. Re. checking links in our docs, the approach that Cam suggested won't work because even after running `pnpm typedoc` and generating the `guide` directory before running the check, the links in the markdown files still contain `v{site.data.versions.sway` etc. These 'phrases' are parsed and replaced _only_ in the `.html` files inside of the `_site` folder.
   
   tldr: to make sure there are no broken links inside of our docs, we should be checking the HTML files generated inside the `_site` folder and _not_ the `.md` files inside of the `guide` folder. 

To summarize:
- We'll need to check the `.html` files inside of `docs/_site` to make sure there are no broken links in our docs. 
- We'll need to check the rest of our `.md` files for any broken links. (these would mostly be `readme` files)

**Comment by Dhaiwat10** [comment]
Marking this as blocked for now

**Comment by arboleya** [comment]
@Dhaiwat10 We might be able to resume this PR after #811 is merged.

> cc @Torres-ssf

**Comment by arboleya** [comment]
@Dhaiwat10 I feel we can resume this PR with the new docs, no?

> cc @Torres-ssf

**Comment by Dhaiwat10** [comment]
Oh yes, we can resume this one now. Will resume work on this 👍🏻  @camsjams @arboleya

**Comment by Dhaiwat10** [comment]
Closing this in favor of #1010
